common/achievement/index.ts

```diff
@@ -1,50 +1,57 @@
 import { prisma } from '../prisma';
 import { User } from '../user';
-import { EventValue } from './types';
-import { getMetricsByAction, getRelationByContext } from './util';
+import { isGamificationFeatureActive, getMetricsByAction } from './util';
 import { getLogger } from '../logger/logger';
 import { ActionID, SpecificNotificationContext } from '../notification/actions';
-import { NotificationContext } from '../notification/types';
-import { Achievement_event } from '../../graphql/generated';
-import { doesTemplateExistForAction } from './template';
+import { getTemplatesByAction } from './template';
+import { evaluateAchievement } from './evaluate';
+import { AchievementToCheck, ActionEvent, ConditionDataAggregations, UserAchievementContext, UserAchievementTemplate } from './types';
+import { createAchievement, getOrCreateUserAchievement } from './create';
+import { Achievement_template } from '../../graphql/generated';
+import { Prisma } from '@prisma/client';
+import { injectRecordValue, sortActionTemplatesToGroups } from './helper';
 
 const logger = getLogger('Achievement');
 
-export type ActionEvent = {
-    actionId: ActionID;
-    at: Date;
-    user: User;
-};
-export type Achievement_Event = {
-    userId?: string;
-    metric: string;
-    value: EventValue;
-    action?: string;
-    relation?: string; // e.g. "user/10", "subcourse/15", "match/20"
-};
 export async function actionTaken<ID extends ActionID>(user: User, actionId: ID, context: SpecificNotificationContext<ID>) {
-    const templateExists = await doesTemplateExistForAction(actionId);
+    if (!isGamificationFeatureActive()) {
+        return;
+    }
+    const templatesForAction = await getTemplatesByAction(actionId);
 
-    if (!templateExists) {
+    if (templatesForAction.length === 0) {
         logger.debug(`No achievement found for action '${actionId}'`);
         return;
     }
 
-    const event: ActionEvent = {
+    const templatesByGroups = sortActionTemplatesToGroups(templatesForAction);
+
+    const event: ActionEvent<ID> = {
         actionId,
         at: new Date(),
         user: user,
+        context,
     };
+    await trackEvent(event);
 
-    const tracked = await trackEvent(event, context);
-    // TODO: check if user achievement already exists
-    // TODO: track event(event)
-    // TODO: checkAwardAchievement
+    for (const [key, group] of templatesByGroups) {
+        let achievementToCheck: AchievementToCheck;
+        for (const template of group) {
+            const userAchievement = await getOrCreateUserAchievement(template, user.userID, context);
+            if (userAchievement.achievedAt === null || userAchievement.recordValue) {
```

Comment:
Why are we using the recordValue to identify the type of an achievement if we have a achievement type field?
```suggestion
            if (userAchievement.achievedAt === null || userAchievement.type === 'STREAK') {
```